pkg/registry/core/pod/strategy.go

```diff
@@ -646,74 +643,3 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
 
 	return container, nil
 }
-
-// applySeccompVersionSkew implements the version skew behavior described in:
-// https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/135-seccomp#version-skew-strategy
-func applySeccompVersionSkew(pod *api.Pod) {
```

Comment:
@tallclair do you have thoughts on the best way to phase this out?

the two audiences I'm considering are:
* manifest authors / API clients
* security scanners / policy enforcers / admission webhooks

ways we could phase out annotation support (unordered):
* stop copying annotations to fields
* stop copying fields to annotations
* stop requiring fields and annotations to match
* reject things that set annotations and not fields
* stop validating seccomp annotations completely

a quick sweep of manifests using the alpha annotations shows that most consumers in github have not backfilled their manifests to include seccomp fields (including manifests in kubernetes and kubernetes-sigs orgs)

I'd be curious to know whether policy enforcers are inspecting both annotations and fields, or just annotations, or just fields. if just annotations, what would the consequences be of letting those enforcers think (based on an annotation) that seccomp was in use when in fact (based on the absence of a seccomp field or a non-matching field) that seccomp policy was not in use?

Reply:
@saschagrunert what's the status of seccomp-by-default? From https://github.com/kubernetes/enhancements/pull/3240 it looks like we're planning to bring it to beta in v1.25, which I assume means default-enabled? If so, I think that mitigates a lot of jordan's concerns about non-updated manifests, since I expect the vast majority of those are setting the default profile.

Here's some numbers from searching public repos:
- `runtime/default`: ~1.8K matches https://sourcegraph.com/search?q=context:global+seccomp%5C.security%5C.alpha%5C.kubernetes%5C.io.*runtime/default+count:2000&patternType=regexp
- `docker/default`: ~5.1K matches https://sourcegraph.com/search?q=context:global+seccomp%5C.security%5C.alpha%5C.kubernetes%5C.io.*docker/default+count:10000&patternType=regexp
- `unconfined`: 663 matches https://sourcegraph.com/search?q=context:global+seccomp%5C.security%5C.alpha%5C.kubernetes%5C.io.*unconfined+count:10000&patternType=regexp
- `localhost`: 258 matches https://sourcegraph.com/search?q=context:global+seccomp%5C.security%5C.alpha%5C.kubernetes%5C.io.*localhost/+count:10000+-repo:kubernetes/kubernetes&patternType=regexp

IMO as long as this is paired with seccomp-by-default, an action required release note should be sufficient to roll this out. One other thing we could consider is updating the [baseline pod-security check](https://github.com/kubernetes/kubernetes/blob/cc6619857027c634918a6d2cca9bab072c46f71f/staging/src/k8s.io/pod-security-admission/policy/check_seccompProfile_baseline.go#L29-L48) to forbid the use of any seccomp annotations.

Reply:
> @saschagrunert what's the status of seccomp-by-default? From [kubernetes/enhancements#3240](https://github.com/kubernetes/enhancements/pull/3240) it looks like we're planning to bring it to beta in v1.25, which I assume means default-enabled?

Since Kubernetes v1.24 beta features are not enabled by default any more. I'll reflect that in the KEP, but still aim to graduate the feature to beta in v1.25 to make it API aware.

Reply:
Ah, I missed that it's opt-in. That's too bad.

In that case, I think for v1.25 we should:
- Stop copying fields to annotations on create
- Warn when setting the annotation

And then push this change out a couple releases, at which point I'd be OK with either ignoring annotations (stop copying them to the field) or explicitly forbidding them.

Reply:
Also the warning should probably be applied to workload resources too if we want anyone to see it.

Another thing we could do: update the pod security restricted profile to forbid setting the seccomp annotations in v1.25+

Reply:
> I think for v1.25 we should:
>
> * Stop copying fields to annotations on create
> * Warn when setting the annotation
>
> And then push this change out a couple releases, at which point I'd be OK with either ignoring annotations (stop copying them to the field) or explicitly forbidding them.

That seems like the right direction to wean people off annotations.

I'd probably keep the validation that says if field and annotation are both set, they have to match, but gradually remove the auto-copying between them (remove field → annotation copying in 1.25, remove annotation → field copying in 1.X)

> Also the warning should probably be applied to workload resources too if we want anyone to see it.

any warning we put in GetWarningsForPod will surface for built-in workloads containing pod specs

> Another thing we could do: update the pod security restricted profile to forbid setting the seccomp annotations in v1.25+

Mmm... that feels sort of weird. I'd consider keeping the validation that says if you specify both, they have to match.

Reply:
> remove annotation → field copying in 1.X

I don't think this is worth doing. I don't want anyone looking at annotations, so I think at the point we stop copying annotations to fields we should consider annotations done, and remove the annotation code entirely.

I agree that we should keep the validation that says if both are set they must match.

Reply:
>> remove annotation → field copying in 1.X
>
> I don't think this is worth doing

Did I misread this:

> And then push this change out a couple releases, at which point I'd be OK with either ignoring annotations (stop copying them to the field)...

I thought that's what you were suggesting

Reply:
Ah, yeah, I think I meant ignoring them holistically, basically what this PR is currently doing.

Reply:
So I guess we're fine here, right?

Reply:
My suggestion was:

> In that case, I think for v1.25 we should:
> - Stop copying fields to annotations on create
> - Warn when setting the annotation

And then push the changes in this PR (completely ignoring annotations) out a release or more.

Reply:
> My suggestion was:
> 
> > In that case, I think for v1.25 we should:
> > 
> > * Stop copying fields to annotations on create
> > * Warn when setting the annotation
> 
> And then push the changes in this PR (completely ignoring annotations) out a release or more.

@liggitt are you fine with that?

Reply:
> In that case, I think for v1.25 we should:
>
> 1. Stop copying fields to annotations on create
> 2. Warn when setting the annotation

That sounds good for 1.25.

> And then push the changes in this PR (completely ignoring annotations) out a release or more.

I'm not sure we should ever drop the validation that forbids a mismatch between annotation and field if both are set. I think it's reasonable to drop the annotation → field copying after a bit.